src/dataregistry/db_basic.py

```diff
@@ -147,96 +156,125 @@ def dialect(self):
-        Keep and dispense table metadata
+        return self._prod_schema
 
-        Parameters
-        ----------
-        db_connection : DbConnection object
-            Stores information about the DB connection
-        get_db_version : bool, optional
-            True to extract the DB version from the provenance table
+    @property
+    def active_schema(self):
+        if self._production_mode:
+            return self._prod_schema
+        else:
+            return self._schema
+
+    @property
+    def production_mode(self):
+        return self._production_mode
+
+    def _reflect(self):
         """
+        Reflect the working and production schemas to get the tables within the database.
 
-        self._metadata = MetaData(schema=db_connection.schema)
-        self._engine = db_connection.engine
-        self._schema = db_connection.schema
+        The production schema is automatically derived from the working schema
+        through the provenance table. The tables and versions of each schema
+        are extracted and stored in the `self.metadata` dict.
 
-        # Load all existing tables
-        self._metadata.reflect(self._engine, db_connection.schema)
+        Note during schema creating the provenance information will not yet be
+        avaliable, hense the warning rather than an exception.
+        """
+        # Reflect the working schema to find database tables
+        metadata = MetaData(schema=self.schema)
+        metadata.reflect(self.engine, self.schema)
 
-        # Fetch and save db versioning, assoc. production schema
-        # if present and requested
-        self._prod_schema = None
-        if db_connection.dialect == "sqlite":
+        # Find the provenance table in the working schema
+        if self.dialect == "sqlite":
             prov_name = "provenance"
         else:
-            prov_name = ".".join([self._schema, "provenance"])
+            prov_name = ".".join([self.schema, "provenance"])
 
-        if prov_name not in self._metadata.tables:
+        if prov_name not in metadata.tables:
             raise DataRegistryException(
                 f"Incompatible database: no Provenance table {prov_name}, "
                 f"listed tables are {self._metadata.tables}"
                 )
 
-        if get_db_version:
-            prov_table = self._metadata.tables[prov_name]
-            stmt = select(column("associated_production")).select_from(prov_table)
+        # From the procenance table get the associated production schema
+        cols = ["db_version_major", "db_version_minor", "db_version_patch", "associated_production"]
+        prov_table = metadata.tables[prov_name]
+        stmt = select(*[column(c) for c in cols]).select_from(prov_table)
+        stmt = stmt.order_by(prov_table.c.provenance_id.desc())
+        with self.engine.connect() as conn:
+            results = conn.execute(stmt)
+            r = results.fetchone()
+        if r is None:
+            warnings.warn(
```

Comment:
I've added a `creation_mode` flag for the `DbConnection` which skips querying the provenance table during reflection. Now if nothing is found in the provenance table, an exception is raised rather than a warning.